### PR TITLE
dark mode: add feature flag and UI supporter

### PR DESCRIPTION
**Pull request by stephanwlee**
This change is the first of many to support dark mode. There is the
blessed way of supporting dark mode in Angular; their theming system
supports it quite well. However, there are large gap between what
Angular and Angular TensorBoard can support vs. what Polymer based
components can support. Until we figure it out, we want to keep it under
wraps by disabling it for our users.

Also, by leveraging the feature flagging with the URL deserialization,
we can use query parameter to force enable the dark mode for testing
easily.

In subsequent changes, we will piece by piece support dark mode
manually.

### CODE REVIEW

File: tensorboard/webapp/core/views/dark_mode_supporter_test.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule],
+      providers: [provideMockStore(), DarkModeSupportContainer],
+      declarations: [DarkModeSupportContainer],
+    }).compileComponents();
+    store = TestBed.inject<Store<State>>(Store) as MockStore<State>;
+    store.overrideSelector(getDarkModeEnabled, false);
+  });
```
Comment (psybuzz) [inline tensorboard/webapp/core/views/dark_mode_supporter_test.ts:35]: Should we need to reset the `document.body.classList` to default in an `afterEach` to not leak state?

File: tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source_test.ts [hunk 2]
```diff
@@ -108,3 +148,6 @@ describe('tb_feature_flag_data_source', () => {
     });
   });
 });
+function DARK_MODE_MEDIA_QUERY(DARK_MODE_MEDIA_QUERY: any) {
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source_test.ts:151]: unused?

File: tensorboard/webapp/webapp_data_source/tb_feature_flag_module.ts [hunk 3]
```diff
@@ -15,15 +15,35 @@ limitations under the License.
 
 import {NgModule} from '@angular/core';
 
-import {TBFeatureFlagDataSource} from './tb_feature_flag_data_source_types';
-import {QueryParamsFeatureFlagDataSource} from './tb_feature_flag_data_source';
+import {
+  TBFeatureFlagDataSource,
+  TbFeatureFlagDataSources,
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_feature_flag_module.ts:20]: Can we use consistent capitalization of `TB` vs `Tb`?

**Review by psybuzz — COMMENTED** [review]
Overall, things look good, except I'm opposed to the introduction of MediaQueryFeatureFlagDataSource.  I think the specialization of FeatureFlagDataSource into a MediaQuery* and QueryParam* variant adds too much complexity at the moment.  A multi-provider requires introduces a more ambiguity about where a particular feature flag is provided, and all TB surfaces need to include this "QueryParameter provider should appear the last" comment.

imo, the only thing we gain by having this split is that other TB surfaces can opt to include only queryParam flags without including the mediaQuery ones, in a way that does not require composition or redefining all the mediaQuery flags.  This benefit only applies if the other TB surface has no other need to compose/redefine feature flag data source.

If we renamed the existing `QueryParamFeatureFlagDataSource` to be `OSSFeatureFlagDataSource` or something, it would be reasonable for it to read from both query params and media queries.  Other TB surfaces might have to compose/redefine their own feature flag data source, but this is something that other TB surfaces will do already.

Alternatively, wdyt about having DarkModeSupportContainer be responsible for calling `window.matchMedia`?  I don't know if the media query result can change over the lifetime of the page session (if user changes the OS setting without refreshing the tab), but if so, maybe we don't need MediaQueryFeatureFlagDataSource at all?

**Comment by stephanwlee** [comment]
> Alternatively, wdyt about having DarkModeSupportContainer be responsible for calling window.matchMedia? I don't know if the media query result can change over the lifetime of the page session (if user changes the OS setting without refreshing the tab), but if so, maybe we don't need MediaQueryFeatureFlagDataSource at all?

It can change over time (OS nowadays changes modes based on time of the day) which I will not support for now but I want a user to be able to toggle it with a button so the state will not only come from the media query. For a stateful things, I really wish we'd use the standard pattern instead of doing some hybrid where convenient.

File: tensorboard/webapp/webapp_data_source/tb_feature_flag_testing.ts [hunk 4]
```diff
@@ -18,6 +18,8 @@ import {Injectable, NgModule} from '@angular/core';
 import {buildFeatureFlag} from '../feature_flag/testing';
 import {TBFeatureFlagDataSource} from './tb_feature_flag_data_source_types';
 
+export {TBFeatureFlagDataSource} from './tb_feature_flag_data_source_types';
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_feature_flag_testing.ts:21]: unneeded/unused?

File: tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source.ts [hunk 5]
```diff
@@ -59,12 +65,34 @@ export class QueryParamsFeatureFlagDataSource extends TBFeatureFlagDataSource {
...
   }
 
   protected getParams() {
     return util.getParams();
   }
+
+  protected getPartialFeaturesFromMediaQuery(): {enableDarkMode?: boolean} {
+    // Set feature flag value for query parameters that are explicitly
+    // specified. Feature flags for unspecified query parameters remain unset so
+    // their values in the underlying state are not inadvertently changed.
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source.ts:84]: Let's update this comment still refers to "query parameters" although the code below reads media queries.  Optionally, I think the comment could also be dropped, given the comment below on L88

File: tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source_types.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-import {Injectable} from '@angular/core';
+import {Injectable, InjectionToken} from '@angular/core';
```
Comment (psybuzz) [inline tensorboard/webapp/webapp_data_source/tb_feature_flag_data_source_types.ts:15]: unused `InjectionToken`

**Review by psybuzz — APPROVED** [review]
Thank you for updating the approach, looking forward to this well requested feature!  https://github.com/tensorflow/tensorboard/issues/1277

**Pull request merged**